compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -290,14 +302,22 @@ pub fn remove_duplicate_unreachable_blocks<'tcx>(tcx: TyCtxt<'tcx>, body: &mut B
         }
 
         fn visit_terminator(&mut self, terminator: &mut Terminator<'tcx>, location: Location) {
+            let mut optimization_was_applied = false;
             for target in terminator.successors_mut() {
                 // We don't have to check whether `target` is a cleanup block, because have
                 // entirely excluded cleanup blocks in building the set of duplicates.
                 if self.duplicates.contains(target) {
+                    if *target != self.duplicates[0] {
+                        optimization_was_applied = true;
+                    }
                     *target = self.duplicates[0];
                 }
             }
 
+            if optimization_was_applied {
```

Comment:
Is this heavy enough to warrant the short circuit?
Could this cause us to miss cases where combining would be useful, but the branches are already deduplicated?
`match { A => {}, B | C => unreachable }`

Reply:
Or do you rely on the call in instcombine?

Reply:
When I wrote this I think I didn't realize how rare the combining should be. And I'm quite sure I do not want to be relying on InstCombine.